File: app/src/main/java/eu/kanade/tachiyomi/ui/catalogue/CataloguePresenter.java
package eu.kanade.tachiyomi.ui.catalogue;

import android.os.Bundle;
import android.text.TextUtils;

import com.pushtorefresh.storio.sqlite.operations.put.PutResult;

import java.util.List;

import javax.inject.Inject;

import eu.kanade.tachiyomi.data.cache.CoverCache;
import eu.kanade.tachiyomi.data.database.DatabaseHelper;
import eu.kanade.tachiyomi.data.database.models.Manga;
import eu.kanade.tachiyomi.data.preference.PreferencesHelper;
import eu.kanade.tachiyomi.data.source.SourceManager;
import eu.kanade.tachiyomi.data.source.base.Source;
import eu.kanade.tachiyomi.data.source.model.MangasPage;
import eu.kanade.tachiyomi.ui.base.presenter.BasePresenter;
import eu.kanade.tachiyomi.util.RxPager;
import icepick.State;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subjects.PublishSubject;
import timber.log.Timber;

public class CataloguePresenter extends BasePresenter<CatalogueFragment> {

    @Inject SourceManager sourceManager;
    @Inject DatabaseHelper db;
    @Inject CoverCache coverCache;
    @Inject PreferencesHelper prefs;

    private List<Source> sources;
    private Source source;
    @State int sourceId;

    private String query;

    private RxPager<Manga> pager;
    private MangasPage lastMangasPage;

    private PublishSubject<List<Manga>> mangaDetailSubject;

    private boolean isListMode;

    private static final int GET_MANGA_LIST = 1;
    private static final int GET_MANGA_DETAIL = 2;
    private static final int GET_MANGA_PAGE = 3;

    @Override
    protected void onCreate(Bundle savedState) {
        super.onCreate(savedState);

        if (savedState != null) {
            source = sourceManager.get(sourceId);
        }

        sources = sourceManager.getSources();

        mangaDetailSubject = PublishSubject.create();

        pager = new RxPager<>();

        startableReplay(GET_MANGA_LIST,
                pager::results,
                (view, pair) -> view.onAddPage(pair.first, pair.second));

        startableFirst(GET_MANGA_PAGE,
                () -> pager.request(page -> getMangasPageObservable(page + 1)),
                (view, next) -> {},
                (view, error) -> view.onAddPageError());

        startableLatestCache(GET_MANGA_DETAIL,
                () -> mangaDetailSubject
                        .observeOn(Schedulers.io())
                        .flatMap(Observable::from)
                        .filter(manga -> !manga.initialized)
                        .concatMap(this::getMangaDetails)
                        .onBackpressureBuffer()
                        .observeOn(AndroidSchedulers.mainThread()),
                CatalogueFragment::updateImage,
                (view, error) -> Timber.e(error.getMessage()));

        add(prefs.catalogueAsList().asObservable()
                .subscribe(this::setDisplayMode));
    }

    private void setDisplayMode(boolean asList) {
        this.isListMode = asList;
        if (asList) {
            stop(GET_MANGA_DETAIL);
        } else {
            start(GET_MANGA_DETAIL);
        }
    }

    public void startRequesting(Source source) {
        this.source = source;
        sourceId = source.getId();
        restartRequest(null);
    }

    public void restartRequest(String query) {
        this.query = query;
        stop(GET_MANGA_PAGE);
        lastMangasPage = null;

        if (!isListMode) {
            start(GET_MANGA_DETAIL);
        }
        start(GET_MANGA_LIST);
        start(GET_MANGA_PAGE);
    }

    public void requestNext() {
        if (hasNextPage()) {
            start(GET_MANGA_PAGE);
        }
    }

    private Observable<List<Manga>> getMangasPageObservable(int page) {
        MangasPage nextMangasPage = new MangasPage(page);
        if (page != 1) {
            nextMangasPage.url = lastMangasPage.nextPageUrl;
        }

        Observable<MangasPage> obs = !TextUtils.isEmpty(query) ?
            source.searchMangasFromNetwork(nextMangasPage, query) :
            source.pullPopularMangasFromNetwork(nextMangasPage);

        return obs.subscribeOn(Schedulers.io())
                .doOnNext(mangasPage -> lastMangasPage = mangasPage)
                .flatMap(mangasPage -> Observable.from(mangasPage.mangas))
                .map(this::networkToLocalManga)
                .toList()
                .doOnNext(this::initializeMangas)
                .observeOn(AndroidSchedulers.mainThread());
    }

    private Manga networkToLocalManga(Manga networkManga) {
        Manga localManga = db.getManga(networkManga.url, source.getId()).executeAsBlocking();
        if (localManga == null) {
            PutResult result = db.insertManga(networkManga).executeAsBlocking();
            networkManga.id = result.insertedId();
            localManga = networkManga;
        }
        return localManga;
    }

    public void initializeMangas(List<Manga> mangas) {
        mangaDetailSubject.onNext(mangas);
    }

    private Observable<Manga> getMangaDetails(final Manga manga) {
        return source.pullMangaFromNetwork(manga.url)
                .flatMap(networkManga -> {
                    manga.copyFrom(networkManga);
                    db.insertManga(manga).executeAsBlocking();
                    return Observable.just(manga);
                })
                .onErrorResumeNext(error -> Observable.just(manga));
    }

    public Source getSource() {
        return source;
    }

    public boolean hasNextPage() {
        return lastMangasPage != null && lastMangasPage.nextPageUrl != null;
    }

    public int getLastUsedSourceIndex() {
        int index = prefs.lastUsedCatalogueSource().get();
        if (index < 0 || index >= sources.size() || !isValidSource(sources.get(index))) {
            return findFirstValidSource();
        }
        return index;
    }

    public boolean isValidSource(Source source) {
        if (!source.isLoginRequired() || source.isLogged())
            return true;

        return !(prefs.getSourceUsername(source).equals("")
                || prefs.getSourcePassword(source).equals(""));
    }

    public int findFirstValidSource() {
        for (int i = 0; i < sources.size(); i++) {
            if (isValidSource(sources.get(i))) {
                return i;
            }
        }
        return 0;
    }

    public void setEnabledSource(int index) {
        prefs.lastUsedCatalogueSource().set(index);
    }

    public List<Source> getEnabledSources() {
        // TODO filter by enabled source
        return sourceManager.getSources();
    }

    public void changeMangaFavorite(Manga manga) {
        manga.favorite = !manga.favorite;
        db.insertManga(manga).executeAsBlocking();
    }

    public boolean isListMode() {
        return isListMode;
    }

    public void swapDisplayMode() {
        prefs.catalogueAsList().set(!isListMode);
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/catalogue/CataloguePresenter.kt
package eu.kanade.tachiyomi.ui.catalogue

import android.os.Bundle
import eu.kanade.tachiyomi.data.preference.PreferencesHelper
import eu.kanade.tachiyomi.data.preference.getOrDefault
import eu.kanade.tachiyomi.source.CatalogueSource
import eu.kanade.tachiyomi.source.LocalSource
import eu.kanade.tachiyomi.source.SourceManager
import eu.kanade.tachiyomi.ui.base.presenter.BasePresenter
import rx.Observable
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import uy.kohesive.injekt.Injekt
import uy.kohesive.injekt.api.get
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Presenter of [CatalogueController]
 * Function calls should be done from here. UI calls should be done from the controller.
 *
 * @param sourceManager manages the different sources.
 * @param preferences application preferences.
 */
class CataloguePresenter(
        val sourceManager: SourceManager = Injekt.get(),
        private val preferences: PreferencesHelper = Injekt.get()
) : BasePresenter<CatalogueController>() {

    /**
     * Enabled sources.
     */
    var sources = getEnabledSources()

    /**
     * Subscription for retrieving enabled sources.
     */
    private var sourceSubscription: Subscription? = null

    override fun onCreate(savedState: Bundle?) {
        super.onCreate(savedState)

        // Load enabled and last used sources
        loadSources()
        loadLastUsedSource()
    }

    /**
     * Unsubscribe and create a new subscription to fetch enabled sources.
     */
    fun loadSources() {
        sourceSubscription?.unsubscribe()

        val map = TreeMap<String, MutableList<CatalogueSource>> { d1, d2 ->
            // Catalogues without a lang defined will be placed at the end
            when {
                d1 == "" && d2 != "" -> 1
                d2 == "" && d1 != "" -> -1
                else -> d1.compareTo(d2)
            }
        }
        val byLang = sources.groupByTo(map, { it.lang })
        val sourceItems = byLang.flatMap {
            val langItem = LangItem(it.key)
            it.value.map { source -> SourceItem(source, langItem) }
        }

        sourceSubscription = Observable.just(sourceItems)
                .subscribeLatestCache(CatalogueController::setSources)
    }

    private fun loadLastUsedSource() {
        val sharedObs = preferences.lastUsedCatalogueSource().asObservable().share()

        // Emit the first item immediately but delay subsequent emissions by 500ms.
        Observable.merge(
                sharedObs.take(1),
                sharedObs.skip(1).delay(500, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread()))
                .distinctUntilChanged()
                .map { (sourceManager.get(it) as? CatalogueSource)?.let { SourceItem(it) } }
                .subscribeLatestCache(CatalogueController::setLastUsedSource)
    }

    fun updateSources() {
        sources = getEnabledSources()
        loadSources()
    }

    /**
     * Returns a list of enabled sources ordered by language and name.
     *
     * @return list containing enabled sources.
     */
    private fun getEnabledSources(): List<CatalogueSource> {
        val languages = preferences.enabledLanguages().getOrDefault()
        val hiddenCatalogues = preferences.hiddenCatalogues().getOrDefault()

        return sourceManager.getCatalogueSources()
                .filter { it.lang in languages }
                .filterNot { it.id.toString() in hiddenCatalogues }
                .sortedBy { "(${it.lang}) ${it.name}" } +
                sourceManager.get(LocalSource.ID) as LocalSource
    }
}
